src/java/org/apache/cassandra/index/sai/memory/TrieMemoryIndex.java

```diff
@@ -129,27 +177,71 @@ public synchronized void add(DecoratedKey key,
             final long initialSizeOffHeap = data.usedSizeOffHeap();
             final long reducerHeapSize = primaryKeysReducer.heapAllocations();
 
+            if (docLengths.containsKey(primaryKey))
+            {
+                AtomicLong heapReclaimed = new AtomicLong();
+                // we're overwriting an existing cell, clear out the old term counts
+                for (Map.Entry<ByteComparable, PrimaryKeys> entry : data.entrySet())
+                {
+                    var termInTrie = entry.getKey();
+                    entry.getValue().forEach(pkInTrie -> {
+                        if (pkInTrie.equals(primaryKey))
+                        {
+                            var t = new PkWithTerm(pkInTrie, termInTrie);
+                            termFrequencies.remove(t);
+                            heapReclaimed.addAndGet(RamUsageEstimator.HASHTABLE_RAM_BYTES_PER_ENTRY + t.ramBytesUsed() + Integer.BYTES);
+                        }
+                    });
+                }
+            }
```

Comment:
Iterating over the whole trie seems expensive. Can we limit our usage of `docLengths` to when the index is analyzed?